### ISSUE TITLE
systemd-resolved.service stops after some time

### ISSUE THREAD

**Issue by Ci7rix** [coreos/fedora-coreos-tracker#1583]
### Describe the bug

After some time, the dns resolution doesn't work anymore with the systemd-resolved sometimes on the whole cluster but not at the same time.
Kubernetes is installed on this node with Calico using Kubespray. 

```
$ dig google.com

; <<>> DiG 9.18.17 <<>> google.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 15627
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 65494
;; QUESTION SECTION:
;google.com.                    IN      A

;; Query time: 0 msec
;; SERVER: 127.0.0.53#53(127.0.0.53) (UDP)
;; WHEN: Mon Sep 25 16:26:33 UTC 2023
;; MSG SIZE  rcvd: 39

$ dig google.com @8.8.8.8

; <<>> DiG 9.18.17 <<>> google.com @8.8.8.8
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 25010
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;google.com.                    IN      A

;; ANSWER SECTION:
google.com.             3       IN      A       172.217.168.14

;; Query time: 6 msec
;; SERVER: 8.8.8.8#53(8.8.8.8) (UDP)
;; WHEN: Mon Sep 25 16:26:34 UTC 2023
;; MSG SIZE  rcvd: 55
```

The log of systemd-resolved 

```
Sep 25 16:29:06 worker02.xxx.yy systemd[1]: Stopping systemd-resolved.service - Network Name Resolution...
Sep 25 16:29:06 worker02.xxx.yy systemd[1]: systemd-resolved.service: Deactivated successfully.
Sep 25 16:29:06 worker02.xxx.yy systemd[1]: Stopped systemd-resolved.service - Network Name Resolution.
Sep 25 16:29:06 worker02.xxx.yy systemd[1]: Starting systemd-resolved.service - Network Name Resolution...
Sep 25 16:29:06 worker02.xxx.yy systemd-resolved[173596]: Positive Trust Anchors:
Sep 25 16:29:06 worker02.xxx.yy systemd-resolved[173596]: . IN DS 20326 8 2 e06d44b80b8f1d39a95c0b0d7c65d08458e880409bbc683457104237c7f8ec8d
Sep 25 16:29:06 worker02.xxx.yy systemd-resolved[173596]: Negative trust anchors: home.arpa 10.in-addr.arpa 16.172.in-addr.arpa 17.172.in-addr.arpa 18.172.in-addr.arpa 19.172.in-addr.arpa 20.172.in-addr.arpa 21.172.in-addr.arpa 22.172.in-addr.arpa 23.172.in-addr.arpa 24.172.in-addr.arpa 25.172.in-addr.arpa 26.172.in-addr.arpa 27.172.in-addr.arpa 28.172.in-addr.arpa 29.172.in-addr.arpa 30.172.in-addr.arpa 31.172.in-addr.arpa 168.192.in-addr.arpa d.f.ip6.arpa corp home internal intranet lan local private test
Sep 25 16:29:06 worker02.xxx.yy systemd-resolved[173596]: Using system hostname 'worker02.xxx.yy'.
Sep 25 16:29:06 worker02.xxx.yy systemd[1]: Started systemd-resolved.service - Network Name Resolution.
Sep 25 16:29:16 worker02.xxx.yy systemd-resolved[173596]: Using degraded feature set UDP instead of UDP+EDNS0 for DNS server 169.254.25.10.
```

Restarting NetworkManager makes it working again but restarting restarting systemd-resolved does not work 

```
$ sudo systemctl restart NetworkManager
$ dig google.com

; <<>> DiG 9.18.17 <<>> google.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 42
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 65494
;; QUESTION SECTION:
;google.com.                    IN      A

;; ANSWER SECTION:
google.com.             36      IN      A       142.250.203.110

;; Query time: 7 msec
;; SERVER: 127.0.0.53#53(127.0.0.53) (UDP)
;; WHEN: Mon Sep 25 16:31:21 UTC 2023
;; MSG SIZE  rcvd: 55
```

### Reproduction steps

1. Installing CoreOS
2. Installing Kubernetes with Kubespray and Calico as CNI
3. Wait... maybe 20 hours

### Expected behavior

  systemd-resolved should resolve "google.com"

### Actual behavior

systemd-resolved does not resolve "google.com"

### System details

- Bare metal - x86_64
- Seems to maybe since the last stable update, I hadn't this issue before

```
$ rpm-ostree status -b
State: idle
AutomaticUpdatesDriver: Zincati
  DriverState: active; periodically polling for updates (last checked Mon 2023-09-25 16:34:24 UTC)
BootedDeployment:
● fedora:fedora/x86_64/coreos/stable
                  Version: 38.20230902.3.0 (2023-09-18T22:16:48Z)
               BaseCommit: 2c39e9b6a6b26226b626f4daf815d5307a96b2cd57463e94f7c618e312a61980
             GPGSignature: Valid signature by 6A51BBABBA3D5467B6171221809A8D7CEB10B464
          LayeredPackages: conntrack-tools ipset python python3-libselinux
```

### Butane or Ignition config

```yaml
variant: fcos
version: 1.4.0
storage:
  files:
    - path: /etc/NetworkManager/system-connections/${interface}.nmconnection
      mode: 0600
      contents:
        inline: |
          [connection]
          id=${interface}
          type=ethernet
          interface-name=${interface}
          [ipv4]
          address1=${ip}/${prefix},${gateway}
          dhcp-hostname=${hostname}
          dns=${nameserver};
          dns-search=
          may-fail=false
          method=manual
    - path: /etc/hostname
      mode: 0644
      contents:
        inline: |
          ${hostname}
    - path: /etc/sysctl.d/reverse-path-filter.conf
      contents:
        inline: |
          net.ipv4.conf.all.rp_filter=1
    - path: /etc/profile.d/systemd-pager.sh
      mode: 0644
      contents:
        inline: |
          # Tell systemd to not use a pager when printing information
          export SYSTEMD_PAGER=cat
passwd:
  users:
    - name: core
      ssh_authorized_keys:
        - ${ssh_authorized_key}
systemd:
  units:
    - name: getty@tty1.service
      dropins:
        - name: autologin-core.conf
          contents: |
            [Service]
            # Override Execstart in main unit
            ExecStart=
            # Add new Execstart with `-` prefix to ignore failure`
            ExecStart=-/usr/sbin/agetty --autologin core --noclear %I $TERM
```

**Comment by dustymabe** [comment]
> Restarting NetworkManager makes it working again but restarting restarting systemd-resolved does not work

I think this kind of makes sense because (unless you manually configured DNS another way) NetworkManager is the one communicating with systemd-resolved to tell it what nameservers to use.


I'm not really sure what the problem is here. The logs you provided all happen in 10 seconds. What is that log output you pasted showing?

You say this has been working and is a new issue with the most recent update? If you could confirm it continues to work fine on the older version that would be great. Also you can view the diff in the package set on the [release notes](https://fedoraproject.org/coreos/release-notes/?arch=x86_64&stream=stable) page to see if we can start to pinpoint what software change could be the cause.

I do notice that there are a few issues in kubespray related to resolved and the `DNSStubListener` so those might lend some clues?

- https://github.com/kubernetes-sigs/kubespray/pull/9875
- https://github.com/kubernetes-sigs/kubespray/pull/9160

**Comment by Ci7rix** [comment]
> I'm not really sure what the problem is here. The logs you provided all happen in 10 seconds. What is that log output you pasted showing?

Sorry, I should have been more precise about this. The message goes on forever like this until it's NetworkManager is restarted

```
Sep 26 19:02:40 worker01.lan.td2311.ch systemd-resolved[767]: Using degraded feature set TCP instead of UDP for DNS server 169.254.25.10.
Sep 26 19:02:40 worker01.lan.td2311.ch systemd-resolved[767]: Using degraded feature set UDP instead of TCP for DNS server 169.254.25.10.
Sep 26 19:03:03 worker01.lan.td2311.ch systemd-resolved[767]: Using degraded feature set TCP instead of UDP for DNS server 169.254.25.10.
Sep 26 19:03:03 worker01.lan.td2311.ch systemd-resolved[767]: Using degraded feature set UDP instead of TCP for DNS server 169.254.25.10.
Sep 26 19:03:10 worker01.lan.td2311.ch systemd-resolved[767]: Using degraded feature set TCP instead of UDP for DNS server 169.254.25.10.
Sep 26 19:03:10 worker01.lan.td2311.ch systemd-resolved[767]: Using degraded feature set UDP instead of TCP for DNS server 169.254.25.10.
```
Is there anyway to downgrade CoreOS knowing it's a clean install or do I need to reinstall an older version ? If not, I'll make a clean install of the previous version.

Thank you for the links you provided, I'll have a look at it.

Thank you for your answers.

**Comment by dustymabe** [comment]
Odd. I haven't seen that before. I wonder if you DNS server is somehow misbehaving.

**Comment by Ci7rix** [comment]
I thought about that also. But I tried reinstalling CoreOS using my internal DNS, Google DNS and even Quad9 DNS and got the same result. The config file is from the documentation.

**Comment by dustymabe** [comment]
[Searching](https://github.com/systemd/systemd/issues?q=is%3Aissue+is%3Aopen+"Using+degraded+feature+set") upstream systemd seems to yield a few results that might be similar:

- https://github.com/systemd/systemd/issues/21174
- https://github.com/systemd/systemd/issues/21123

I don't think there is any resolution right now. This can't be a super common problem because everyone would be complaining, but obviously there is some issue (maybe some corner case) because there are people reporting it upstream.

**Comment by Ci7rix** [comment]
Thank you for your reply.

I discovered I had the same issue on a Fedora Server 38 cloud-init image. It may be related to my network, I just don't know how as all those servers have static IPs and external DNS server.

I downgraded my whole cluster yesterday and reinstalled Kubernetes, I'll update 2–3 servers and wait to see if the issue happens again.

**Comment by dustymabe** [comment]
> Thank you for your reply.

No problem. I highly suggest you add your data point to the upstream systemd issues. Maybe they can help work out what the problem is for everyone if you understand your network environment well enough and are able to debug with them. 

To get you unblocked, though, you could possibly just disable systemd-resolved.
